src/blas/impl/KokkosBlas3_gemm_impl.hpp

```diff
@@ -0,0 +1,506 @@
+            const int idx_i = offset_i+i;
+            A(idx_i,idx_j) = beta * A(idx_i,idx_j) + alpha * A_scr(i,j);
+          });
+        }
+      });
+    }
+  }
+};
+
+template<class TeamHandle, class ViewType, class ViewTypeScratch, int blockDim_i, int blockDim_j>
+struct impl_update_matrix_block<TeamHandle,ViewType,ViewTypeScratch,Kokkos::LayoutRight,blockDim_i,blockDim_j> {
+  typedef typename ViewType::non_const_value_type value_type;
+  typedef Kokkos::Details::ArithTraits<value_type>     ATV;
+
+  KOKKOS_INLINE_FUNCTION
+  static void update(const TeamHandle& team, const value_type& beta , const ViewType& A,
+                                             const value_type& alpha, const ViewTypeScratch& A_scr,
+                                             const int& offset_i, const int& offset_j) {
+    if(offset_i + blockDim_i <= A.extent(0) && offset_j + blockDim_j <= A.extent(1)) {
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team,blockDim_i), [&] (const int i) {
+        const int idx_i = offset_i+i;
+        if(beta == ATV::zero()) {
+          Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,blockDim_j), [&] (const int j) {
+            const int idx_j = offset_j+j;
+            A(idx_i,idx_j) = alpha * A_scr(i,j);
+          });
+        } else {
+          Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,blockDim_j), [&] (const int j) {
+            const int idx_j = offset_j+j;
+            A(idx_i,idx_j) = beta * A(idx_i,idx_j) + alpha * A_scr(i,j);
+          });
+        }
+      });
+    } else {
+      const int range_i = offset_i + blockDim_i <= A.extent(0)?blockDim_i:A.extent(0)%blockDim_i;
+      const int range_j = offset_j + blockDim_j <= A.extent(1)?blockDim_j:A.extent(1)%blockDim_j;
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team,range_i), [&] (const int i) {
+        const int idx_i = offset_i+i;
+        if(beta == ATV::zero()) {
+          Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,range_j), [&] (const int j) {
+            const int idx_j = offset_j+j;
+            A(idx_i,idx_j) = alpha * A_scr(i,j);
+          });
+        } else {
+          Kokkos::parallel_for(Kokkos::ThreadVectorRange(team,range_j), [&] (const int j) {
+            const int idx_j = offset_j+j;
+            A(idx_i,idx_j) = beta * A(idx_i,idx_j) + alpha * A_scr(i,j);
+          });
+
+        }
+      });
+    }
+  }
+};
+
+// Compute a single A block 8 B block, also do an in-place no-additional blocking team GEMM
+template<class TeamHandle, class ViewTypeA, class ViewTypeB, class ViewTypeC>
+KOKKOS_INLINE_FUNCTION
+void impl_team_gemm_block(const TeamHandle& team, const ViewTypeC& C, const ViewTypeA& A, const ViewTypeB& B) {
+  typedef typename ViewTypeC::non_const_value_type ScalarC;
+  const int blockA0 = A.extent(0);
+  const int blockA1 = A.extent(1);
+  const int blockB1 = B.extent(1);
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team,blockA0), [&] (const int i) {
```

Comment:
Tuning capability between team size and blockA0 size. On a GPU, team size can be determined by cuda function considering functor size and shared memory (max or optimal in their context). This means that blockA0 is derived from team size ? or team size is chosen by either the cuda suggestion or blockA0 ?

Reply:
TeamSize is chose based on blockA0, otherwise you either have idle threads, or use more scratch memory per thread.